### PR TITLE
Refactoring Deploy.ps1 and Publish.ps1 to reuse Write-Log.ps1

### PR DESCRIPTION
**Pull request by adbindal**
## Description
- As a part of creating `Cleanup.ps1`, we refactored logging into a separate utility script of its own. This PR starts re-using that inside `Deploy.ps1` (used via run.bat) and `Publish.ps1` (can be used for only publishing assets to an existing deployment).
- Also, refactored the code inside `Deploy.ps1` to take parameters via command-line as well and moved out exception handling to be more explicit.

## Example with all parameters
``` powershell
Set-ExecutionPolicy -Scope Process -ExecutionPolicy Bypass
.\Deploy.ps1 -ResourceGroupName {rg-name} -AppName {app-name} -IdentityName {id-name} -UseActiveAzureAccount -SubscriptionName {subscription} -LocationName {region}
```

## Testing
- [x] Ensure deployment succeeds with no arguments -- no regression
- [x] Ensure deployment succeeds for custom Resource Group, Identity Name and App Name
- [x] Ensure deployment succeeds when reusing the existing Azure Account (no explicit sign-in)
- [x] Ensure deployment succeeds by passing subscription via command-line (no input other than region)
- [x] Ensure deployment succeeds by passing region via command-line (no input)

### CODE REVIEW

**Comment by adbindal** [comment]
## Code Review Tips
- Look through the changes commit by commit (probably in reverse order), that should isolate the changes to logical entities.
- In case the diff seems quite big, **`Hide whitespace changes`** should be a good option to apply as that might streamline the differences.
image

File: deployment/Deploy.ps1 [hunk 1]
```diff
@@ -102,7 +102,7 @@ process {
         $SubscriptionList = Get-LtiSubscriptionList
         Write-Log -Message "List of Subscriptions:-`n$($SubscriptionList | ConvertTo-Json -Compress)"    
 
-        $SubscriptionCount = $SubscriptionList.Count
+        $SubscriptionCount = ($SubscriptionList | Measure-Object).Count
```
Comment (nehalilani) [inline deployment/Deploy.ps1:110]: why have we included `Measure-Object` here?

File: deployment/Deploy.ps1 [hunk 2]
```diff
@@ -122,7 +122,7 @@ process {
         }
 
         $ActiveSubscription = Set-LtiActiveSubscription -Name $SubscriptionName -List $SubscriptionList
-        $UserEmailAddress = $ActiveSubscription.user.email
+        $UserEmailAddress = $ActiveSubscription.user.name
```
Comment (nehalilani) [inline deployment/Deploy.ps1:130]: why this change?

File: deployment/Deploy.ps1 [hunk 3]
```diff
@@ -3,297 +3,312 @@
...
-    $roleAssignmentOp = az role assignment create --assignee-object-id $identityObj.principalId --assignee-principal-type ServicePrincipal --role $roleName
-    if(!$roleAssignmentOp) {
-        throw "Encountered an Error while creating Role Assignment"
-    }
+        Write-Log -Message "Fetching List of Locations"
+        $LocationList = ((az account list-locations) | ConvertFrom-Json)
+        Write-Log -Message "List of Locations:-`n$($locationList | ConvertTo-Json -Compress)"
 
-    Write-Host 'Role Assignment Created Successfully'
+        if(!$LocationName) {
```
Comment (nehalilani) [inline deployment/Deploy.ps1:140]:
```suggestion
        if($LocationName) {
        Write-Log -Message "Using User provided Location Name: $LocationName"
        }else {
            Write-Host "$(az account list-locations --output table --query "[].{Name:name}" | Out-String)`n"
            $LocationName = Read-Host 'Enter Location From Above List for Resource Provisioning'
            Write-Log -Message "User Entered Location Name: $LocationName"
        }


```

**Review by nehalilani — COMMENTED** [review]
Thank you for the Code Review Tips. 
Changes look good to me. Just a few suggestion/doubts inline

File: deployment/Deploy.ps1 (hunk 1 above)
Comment (adbindal) [reply deployment/Deploy.ps1:110]: The way I understand it is that `$SubscriptionList` is an iterable and without `Measure-Object`, it does not have any objects. Hence, we could either have done the `foreach` stuff that was happening before, or this.

File: deployment/Deploy.ps1 (hunk 2 above)
Comment (adbindal) [reply deployment/Deploy.ps1:130]: Are we talking about the `$UserEmailAddress` or the `Set-LtiActiveSubscription` refactor?
- `$UserEmailAddress` was getting previously set as well but that was happening inside a `foreach` loop when we matched a subscription with a name. I simply separated out the 2 things -- matching subscription and using it to update appropriate variables.
-  The refactor was something which I did with a bit longer term goal of moving every part into its own module (like `Write-Log.ps1`). This should allow us to re-use these across `Deploy`, `Publish` and `Cleanup` scripts.

File: deployment/Deploy.ps1 (hunk 3 above)
Comment (adbindal) [reply deployment/Deploy.ps1:140]: Thanks for the logging suggestion. Committing it right away.

File: deployment/Deploy.ps1 (hunk 2 above)
Comment (nehalilani) [reply deployment/Deploy.ps1:130]: the change I was talking about is user.`email` -> user.`name`

Comment (nehalilani) [reply deployment/Deploy.ps1:130]: I see this is something already present in the deploy script!

**Review by nehalilani — APPROVED** [review]

**Comment by adbindal** [comment]
> Merging changes with 🤞 .
